clients/callback_client.go

```diff
@@ -7,41 +7,39 @@ import (
 	"net/http"
 	"time"
 
+	"github.com/hashicorp/go-retryablehttp"
 	"github.com/livepeer/catalyst-api/config"
 )
 
-const max_retries = 3
-const backoff_millis = 200
-
 type CallbackClient struct {
-	httpClient *http.Client
+	httpClient *retryablehttp.Client
 }
 
 func NewCallbackClient() CallbackClient {
+	client := retryablehttp.NewClient()
+	client.RetryMax = 2                          // Retry a maximum of this+1 times
+	client.RetryWaitMin = 200 * time.Millisecond // Wait at least this long between retries
+	client.RetryWaitMax = 1 * time.Second        // Wait at most this long between retries (exponential backoff)
+	client.HTTPClient = &http.Client{
+		Timeout: 5 * time.Second, // Give up on requests that take more than this long
+	}
+
 	return CallbackClient{
-		httpClient: &http.Client{
-			Timeout: 5 * time.Second,
-		},
+		httpClient: client,
 	}
 }
 
-func (c CallbackClient) DoWithRetries(r *http.Request) error {
-	var resp *http.Response
-	var err error
-	for x := 0; x < max_retries; x++ {
-		resp, err = c.httpClient.Do(r)
-		if err == nil && resp.StatusCode == http.StatusOK {
-			return nil
-		}
-
-		// Back off to give us more chance of succeeding during a network blip
-		time.Sleep(backoff_millis * time.Millisecond)
+func (c CallbackClient) DoWithRetries(r *retryablehttp.Request) error {
+	resp, err := c.httpClient.Do(r)
+	if err != nil {
+		return fmt.Errorf("failed to send callback to %q. Error: %s", r.URL.String(), err)
 	}
 
-	if err != nil {
-		return fmt.Errorf("failed to send callback to %q. Error: %q", r.URL.String(), err)
+	if resp.StatusCode >= 400 {
+		return fmt.Errorf("failed to send callback to %q. HTTP Code: %d", r.URL.String(), resp.StatusCode)
 	}
-	return fmt.Errorf("failed to send callback to %q. Response Status Code: %d", r.URL.String(), resp.StatusCode)
+
+	return nil
 }
```

Comment:
Is this still necessary? I think it could be replaced by a simpler helper like `responseToError(res *http.Response) error`

Reply:
Not sure I 100% understand where the helper would sit, but we do still need the StatusCode check since the retry library only errors on `>= 500`